File: cracking _the_coding_interview/python/面试题 03.01. 三合一.py
class TripleInOne(object):

    def __init__(self, stackSize):
        """
        :type stackSize: int
        """
        self.llen = [0] * 3
        self.vec = [0] * (stackSize * 3)
        self.size = stackSize


    def push(self, stackNum, value):
        """
        :type stackNum: int
        :type value: int
        :rtype: None
        """
        if self.llen[stackNum] >= self.size:
            return
        self.vec[stackNum * self.size + self.llen[stackNum]] = value
        self.llen[stackNum] += 1


    def pop(self, stackNum):
        """
        :type stackNum: int
        :rtype: int
        """
        if self.llen[stackNum] <= 0:
            return -1
        ret = self.vec[stackNum * self.size + self.llen[stackNum] - 1]
        self.llen[stackNum] -= 1
        return ret


    def peek(self, stackNum):
        """
        :type stackNum: int
        :rtype: int
        """
        if self.llen[stackNum] <= 0:
            return -1
        return self.vec[stackNum * self.size + self.llen[stackNum] - 1]


    def isEmpty(self, stackNum):
        """
        :type stackNum: int
        :rtype: bool
        """
        return self.llen[stackNum] == 0


# Your TripleInOne object will be instantiated and called as such:
# obj = TripleInOne(stackSize)
# obj.push(stackNum,value)
# param_2 = obj.pop(stackNum)
# param_3 = obj.peek(stackNum)
# param_4 = obj.isEmpty(stackNum)

File: cracking _the_coding_interview/cpp/面试题 03.01. 三合一.cpp
class TripleInOne {
private:
    int stackSize;
    std::vector<int> len;
    std::vector<int> vec;
public:
    TripleInOne(int stackSize) {
        this->stackSize = stackSize;
        len = std::vector<int>(3, 0);
        vec = std::vector<int>(stackSize * 3, 0);
    }
    
    void push(int stackNum, int value) {
        if (len[stackNum] >= stackSize) {
            return;
        }
        vec[stackNum * stackSize + len[stackNum]] = value;
        len[stackNum] += 1;
    }
    
    int pop(int stackNum) {
        if (len[stackNum] <= 0) {
            return -1;
        }
        int ret = vec[stackNum * stackSize + len[stackNum] - 1];
        len[stackNum] -= 1;
        return ret;
    }
    
    int peek(int stackNum) {
        if (len[stackNum] <= 0) {
            return -1;
        }
        return vec[stackNum * stackSize + len[stackNum] - 1];
    }
    
    bool isEmpty(int stackNum) {
        return len[stackNum] == 0;
    }
};

/**
 * Your TripleInOne object will be instantiated and called as such:
 * TripleInOne* obj = new TripleInOne(stackSize);
 * obj->push(stackNum,value);
 * int param_2 = obj->pop(stackNum);
 * int param_3 = obj->peek(stackNum);
 * bool param_4 = obj->isEmpty(stackNum);
 */